backend/src/main/java/com/google/sps/servlets/SearchMatchServlet.java

```diff
@@ -142,9 +144,45 @@ private void sendMatchResponse(HttpServletResponse response, Match match) throws
     matchExists.put(JSON_FIRST_PARTICIPANT_USERNAME, match.getFirstParticipantUsername());
     matchExists.put(JSON_SECOND_PARTICIPANT_USERNAME, match.getSecondParticipantUsername());
     matchExists.put(JSON_DURATION, match.getDuration());
-
+    sendEmailResponse(true);
     // Send the JSON back as the response
     response.setContentType("application/json");
     response.getWriter().println(matchExists.toString());
   }
+
+  /**
+   * Retrieve user email address via Users API and parse for username. Email can be null when the
+   * current user does not exist.
+   */
+  @Nullable
+  private String getUsername() {
+    String email = userService.getCurrentUser().getEmail();
+    return email != null ? email.split("@")[0] : null;
+  }
+
+  private EmailNotifier createOrGetEmailNotifier() throws GeneralSecurityException, IOException {
+    if (this.emailNotifier != null) {
+      return this.emailNotifier;
+    }
+    this.emailNotifier = new EmailNotifier(gmFactory.build());
+    return this.emailNotifier;
+  }
+
+  private void sendEmailResponse(boolean match) {
+    if (match) {
+      try {
+        emailNotifier = createOrGetEmailNotifier();
+        emailNotifier.sendMatchEmail(getUsername(), userService.getCurrentUser().getEmail());
+      } catch (GeneralSecurityException | MessagingException | IOException e) {
+        e.printStackTrace();
+      }
+    } else {
+      try {
+        emailNotifier = createOrGetEmailNotifier();
+        emailNotifier.sendExpiredEmail(getUsername(), userService.getCurrentUser().getEmail());
+      } catch (GeneralSecurityException | MessagingException | IOException e) {
+        e.printStackTrace();
+      }
+    }
+  }
```

Comment:
Hmmm, when I suggested this I didn't notice that the method calls on emailNotifier are different. Since they are I'd go further and replace this method with two methods, sendMatchEmail and sendExpiredEmail. A method that takes a boolean parameter that is used to switch between two different branches a sign that the method should be split in two.

Reply:
Ack split the function into two